src/Time/StepChoosers/StepChooser.hpp

```diff
@@ -110,7 +103,54 @@ class StepChooser : public PUP::able {
 
   WRAPPED_PUPable_abstract(StepChooser);  // NOLINT
 
-  using creatable_classes = Registration::registrants<StepChooserRegistrars>;
+  /// The `last_step_magnitude` parameter describes the slab size to be
+  /// adjusted; It may be infinite if the appropriate size cannot be determined.
+  ///
+  /// As opposed to the decision made by `desired_step` in the
+  /// StepChooser<StepChooserUse::LtsStep> specialization, the slab
+  /// change decision must be callable from an event (so cannot store
+  /// state information in the \ref DataBoxGroup "DataBox"), and we do
+  /// not have the capability to reject a slab so this function
+  /// returns only a `double` indicating the desired slab size.
+  template <typename Metavariables, typename DbTags>
+  double desired_slab(
+      const double last_step_magnitude,
+      const db::DataBox<DbTags>& box,
+      const Parallel::GlobalCache<Metavariables>& cache) const noexcept {
+    ASSERT(last_step_magnitude > 0.,
+           "Passed non-positive step magnitude: " << last_step_magnitude);
+    using factory_classes =
+        typename std::decay_t<decltype(db::get<Parallel::Tags::Metavariables>(
+            box))>::factory_creation::factory_classes;
+    const auto result =
+        call_with_dynamic_type<std::pair<double, bool>,
+                               tmpl::at<factory_classes, StepChooser>>(
+            this, [&last_step_magnitude, &box,
+                   &cache](const auto* const chooser) noexcept {
```

Comment:
should this return `-> decltype(auto)` or an explicit type, or perhaps make the type of the `result` variable explicit, or use a structured binding?

Reply:
I'm not clear on what problem you are trying to address here.  What do you find bad or unclear?

Responding to a couple of the particular suggestions:

I think clang-tidy will consider `call_with_dynamic_type` a "cast-like" function, so it should get annoyed if I declare the type of `result`.

Structured bindings will make the similar function below awkward (it would disassemble the pair only to immediately reassemble it), and I'd prefer to keep the two functions similar.

Reply:
First I wasn't clear on the return type of this function, and worried it might do a copy because of the implicit `auto` return type. Then I noticed the `std::pair<double, bool>` template parameter and thought perhaps a structured binding would be appropriate. `std::pair`s are so non-descriptive since it's not clear what the `double` and the `bool` mean, in particular in this case where the called function is hidden in a few layers of indirection.

Reply:
Yeah, if you don't like the interface take it up with @moxcodes.

Reply:
@nilsleiffischer The lambda here is intended to be just a forwarding method to call the function in the derived class. The return type for the LTS step choice is detailed in the explanation on line 159 and onward.
 If you think there are better places to move or copy that explanation to, let me know and I'll update that in a separate PR of my own.

Reply:
If you're taking another pass at this interface at some point, I'd suggest to return a struct with named parameters instead of the anonymous std::pair. Not worth the distraction right now though, so don't worry about it :)